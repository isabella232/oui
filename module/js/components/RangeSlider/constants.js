import { amberBase, aquaBase, brandBlueDark, greenBase, orangeBase, pinkBase, redBase, magentaBase, greyBase } from '../../tokens/forimport/index.es'; // map fillColorName prop values to OUI color tokens

export var FILL_COLOR_MAP = {
  aqua: aquaBase,
  amber: amberBase,
  "default": brandBlueDark,
  green: greenBase,
  orange: orangeBase,
  pink: pinkBase,
  red: redBase,
  magenta: magentaBase,
  grey: greyBase
};
export default {
  FILL_COLOR_MAP: FILL_COLOR_MAP
};