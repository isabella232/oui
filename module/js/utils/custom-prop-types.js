import PropTypes from 'prop-types';
export var checkPropIsGreaterThanOrEqualTo = function checkPropIsGreaterThanOrEqualTo(numberToCheck, isRequired) {
  return function (props, propName, componentName) {
    if (isRequired) {
      if (props[propName] === undefined || props[propName] === null) {
        return new Error('Invalid prop `' + propName + '` (has value `' + props[propName] + '`) supplied to `' + componentName + '`. Must be defined.`');
      }
    }

    if (props[propName] < numberToCheck) {
      return new Error('Invalid prop `' + propName + '` (has value `' + props[propName] + '`) supplied to `' + componentName + '`. Must be greater than or equal to `' + numberToCheck + '.');
    }

    for (var _len = arguments.length, rest = new Array(_len > 3 ? _len - 3 : 0), _key = 3; _key < _len; _key++) {
      rest[_key - 3] = arguments[_key];
    }

    return PropTypes.number.apply(PropTypes, [props, propName, componentName].concat(rest));
  };
};