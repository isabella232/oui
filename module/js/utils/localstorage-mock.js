// setup file
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({
  adapter: new Adapter()
});

var localStorageMock = function () {
  var store = {};
  return {
    getItem: function getItem(key) {
      return store[key];
    },
    setItem: function setItem(key, value) {
      store[key] = value.toString();
    },
    clear: function clear() {
      store = {};
    },
    removeItem: function removeItem(key) {
      delete store[key];
    }
  };
}();

Object.defineProperty(window, 'localStorage', {
  value: localStorageMock
});