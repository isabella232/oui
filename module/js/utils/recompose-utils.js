import { withState, withHandlers, compose } from 'recompose';
export var withToggle = compose(withState('isOpen', 'toggle', function (_ref) {
  var isOpen = _ref.isOpen;
  return isOpen || false;
}), withHandlers({
  show: function show(_ref2) {
    var toggle = _ref2.toggle;
    return function (e) {
      return toggle(true);
    };
  },
  hide: function hide(_ref3) {
    var toggle = _ref3.toggle;
    return function (e) {
      return toggle(false);
    };
  },
  toggle: function toggle(_ref4) {
    var _toggle = _ref4.toggle;
    return function (e) {
      return _toggle(function (current) {
        return !current;
      });
    };
  }
}));