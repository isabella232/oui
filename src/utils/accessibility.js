import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import { noop } from 'lodash';

import {
  yellowDark,
  yellowXLight,
  aquaDark,
  aquaXLight,
  brandBlueDark,
  greenDark,
  greenXLight,
  orangeDark,
  orangeXLight,
  pinkDark,
  pinkXLight,
  redDark,
  redXLight,
  magentaDark,
  magentaXLight,
  greyDark,
  blueXLight,
  purpleDark,
  black,
  purpleXLight,
} from '../tokens/forimport/index.es';

/**
 * Map a color class to a human readable word that explains the context of when
 * the class is used. This is helpful for screen readers.
 * @param {String} className - an OUI class
 * @returns {String} English word describing the class provided
 */
export const getAssistiveTextFromColorClass = className => {
  const classMapping = {
    'bad-news': 'Error',
    brand: 'Info',
    'good-news': 'Success',
    warning: 'Alert',
  };

  let text = classMapping[className];

  if (!text) {
    throw new Error('Provided class name does not map to a word.');
  }

  return text;
};

/**
 * HOC that provides support for tracking keyboard actions inside an
 * element with an input. Tracks up & down arrow keys and invokes
 * a callback when enter is pressed.
 * @param {React.Component} Component - The component to wrap.
 * @returns {React.Component}
 */
export const keyboardTracker = Component => {
  const wrappedComponent = props => {
    const { additionalItems, ...rest } = props;
    const [currentFauxFocusIndex, setIndex] = useState(0);
    const [currentItemCount, setItemCount] = useState(0);
    const [onItemSelect, setOnItemSelect] = useState(() => noop);
    /**
     * Event handler for keyboard activity.
     * Increments/decrements the current index on up/down arrow keys
     *  -or-
     * Invokes the onItemSelect callback on enter
     */
    const handleKeyDown = useCallback(
      event => {
        let newIndex;
        switch (event.key) {
          case 'ArrowUp':
            newIndex = Math.max(currentFauxFocusIndex - 1, 0);
            setIndex(newIndex);
            event.preventDefault();
            break;
          case 'ArrowDown':
            const totalItemCount = currentItemCount + additionalItems;
            newIndex = Math.max(Math.min(currentFauxFocusIndex + 1, totalItemCount - 1), 0);
            setIndex(newIndex);
            event.preventDefault();
            break;
          case 'Enter':
            // Select the item for currentFauxFocusIndex
            onItemSelect(currentFauxFocusIndex);
            event.preventDefault();
            break;
          default:
            // No op
            break;
        }
      },
      [currentFauxFocusIndex, currentItemCount, onItemSelect, additionalItems]
    );
    /**
     * Handler to update the item count, which should reset the
     * current index to 0 if the number has changed.
     */
    const handleSetItemCount = useCallback((count) => {
      setItemCount(count);
      if (count !== currentItemCount) {
        setIndex(0);
      }
    }, [currentItemCount]);

    return (
      <Component
        { ...rest }
        currentFauxFocusIndex={ currentFauxFocusIndex }
        handleKeyDown={ handleKeyDown }
        setItemCount={ handleSetItemCount }
        setOnItemSelect={ setOnItemSelect }
      />
    );
  };
  wrappedComponent.displayName = `withkeyboardTracker(${Component.displayName})`;
  wrappedComponent.propTypes = {
    additionalItems: PropTypes.number,
  };
  wrappedComponent.defaultProps = {
    additionalItems: 0,
  };
  return wrappedComponent;
};

// map fillColorName prop values to OUI color tokens
export const FILL_COLOR_MAP = {
  aqua: aquaDark,
  yellow: yellowDark,
  default: brandBlueDark,
  green: greenDark,
  orange: orangeDark,
  pink: pinkDark,
  red: redDark,
  magenta: magentaDark,
  grey: greyDark,
  purple: purpleDark,
  black: black,
};

// map fillColorName prop values to OUI color tokens
export const FILL_COLOR_MAP_LIGHT = {
  aqua: aquaXLight,
  yellow: yellowXLight,
  blue: blueXLight,
  green: greenXLight,
  orange: orangeXLight,
  pink: pinkXLight,
  red: redXLight,
  magenta: magentaXLight,
  purple: purpleXLight,
};

export default {
  FILL_COLOR_MAP,
  FILL_COLOR_MAP_LIGHT,
};
