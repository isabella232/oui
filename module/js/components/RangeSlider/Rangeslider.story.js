import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, number, select, boolean as _boolean } from '@storybook/addon-knobs';
import RangeSlider from './index.js';
import { FILL_COLOR_MAP } from './constants'; // build fillColorName options for a select/dropdown knob

var fillColorOptions = {
  None: ''
};
Object.keys(FILL_COLOR_MAP).forEach(function (color) {
  fillColorOptions[color] = color;
});
var stories = storiesOf('RangeSlider', module);
stories.addDecorator(withKnobs).addDecorator(function (story) {
  return React.createElement("div", {
    id: "root-preview"
  }, story());
});
stories.add('Default', function () {
  return React.createElement(RangeSlider, {
    value: number('value', 50)
  });
}).add('Disabled', function () {
  return React.createElement("div", null, React.createElement("input", {
    type: "text",
    className: "oui-text-input push-double--bottom"
  }), React.createElement(RangeSlider, {
    isDisabled: true,
    value: number('value', 50)
  }), React.createElement("input", {
    type: "text",
    className: "oui-text-input"
  }));
}).add('Custom Fill Color', function () {
  return React.createElement("div", null, React.createElement(RangeSlider, {
    fillColorName: select('fillColorName', fillColorOptions, 'red'),
    value: number('value', 50)
  }));
}).add('Use Off instead of 0%', function () {
  return React.createElement("div", null, React.createElement(RangeSlider, {
    value: number('value', 50),
    useOffForLabel: _boolean('useOffForLabel', true)
  }));
});