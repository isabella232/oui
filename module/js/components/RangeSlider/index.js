import PropTypes from 'prop-types';
import React from 'react';
import classnames from 'classnames';
import { FILL_COLOR_MAP } from './constants';
var DEFAULT_FILL_COLOR_NAME = 'default';

var RangeSlider = function RangeSlider(props) {
  var fillColorName = props.fillColorName,
      value = props.value,
      onChange = props.onChange,
      isDisabled = props.isDisabled;
  var rangeClasses = classnames({
    'oui-rangeslider': true,
    'oui-rangeslider--disabled': isDisabled
  }); // ensure valid fillColor name (in the case that propType errors are ignored)

  var validFillColorName = Object.keys(FILL_COLOR_MAP).includes(fillColorName) ? fillColorName : DEFAULT_FILL_COLOR_NAME;
  var fillColor = FILL_COLOR_MAP[validFillColorName];
  return React.createElement("div", {
    className: rangeClasses,
    "data-test-section": props.testSection
  }, React.createElement("div", {
    className: "oui-grid"
  }, React.createElement("div", {
    className: "oui-grid__cell position--relative"
  }, React.createElement("div", {
    className: "range-labels flex"
  }, React.createElement("label", {
    className: "oui-label muted flush flex--1"
  }, props.useOffForLabel ? 'Off' : '0%'), React.createElement("label", {
    className: "oui-label muted flush"
  }, "100%")), React.createElement("div", {
    className: "range-display"
  }, React.createElement("svg", {
    preserveAspectRatio: "none",
    xmlns: "http://www.w3.org/2000/svg",
    width: "256",
    height: "256",
    viewBox: "0 0 256 256"
  }, React.createElement("path", {
    fill: fillColor,
    d: "M0 256h256V0z"
  })), React.createElement("div", {
    className: "range-display range-display-overlay",
    style: {
      left: value + '%'
    }
  })), React.createElement("div", {
    className: "range-grid"
  }, React.createElement("div", {
    className: "gridline grid-1"
  }), React.createElement("div", {
    className: "gridline grid-2"
  }), React.createElement("div", {
    className: "gridline grid-3"
  }), React.createElement("div", {
    className: "gridline grid-4"
  }), React.createElement("div", {
    className: "gridline grid-5"
  }), React.createElement("div", {
    className: "gridline gridline-half grid-h1"
  }), React.createElement("div", {
    className: "gridline gridline-half grid-h2"
  }), React.createElement("div", {
    className: "gridline gridline-half grid-h3"
  }), React.createElement("div", {
    className: "gridline gridline-half grid-h4"
  })), React.createElement("input", {
    className: "hard position--relative width--1-1",
    "data-traffic-group": "1",
    type: "range",
    min: "0",
    max: "100",
    step: "1",
    value: value,
    onChange: onChange,
    disabled: isDisabled
  }))));
};

RangeSlider.propTypes = {
  /** Custom color for slider **/
  fillColorName: PropTypes.oneOf(Object.keys(FILL_COLOR_MAP)),

  /** Whether it is disabled, will render as greyscale if so **/
  isDisabled: PropTypes.bool,

  /** onChange function */
  onChange: PropTypes.func,

  /** Hook for automated JavaScript tests */
  testSection: PropTypes.string,

  /** When true, uses 'Off' instead of '0%' */
  useOffForLabel: PropTypes.bool,

  /** The value */
  value: PropTypes.number
};
RangeSlider.defaultProps = {
  fillColorName: DEFAULT_FILL_COLOR_NAME,
  isDisabled: false,
  onChange: undefined,
  testSection: undefined,
  useOffForLabel: false,
  value: undefined
};
export default RangeSlider;