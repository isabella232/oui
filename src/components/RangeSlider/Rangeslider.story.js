import React from 'react';

import { storiesOf } from '@storybook/react';
import { withKnobs, number } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';

import RangeSlider from './index.js';

const stories = storiesOf('RangeSlider', module);
stories
  .addDecorator(withKnobs)
  .addDecorator(story => (
    <div id="root-preview">
      {story()}
    </div>
  ));

stories.add('Default', withInfo()(() => {
  return (
    <RangeSlider
      value={ number('value', 50) }
    />
  );
}))
  .add('Disabled', withInfo()(() => {
    return (
      <RangeSlider
        isDisabled={ true }
        value={ number('value', 50) }
      />
    );
  }))
  .add('Test focus state', withInfo()(() => {
    return (
      <div>
        <input type="text" className="oui-text-input push-double--bottom" />
        <RangeSlider
          value={ number('value', 50) }
        />
        <input type="text" className="oui-text-input" />
      </div>
    );
  }));
