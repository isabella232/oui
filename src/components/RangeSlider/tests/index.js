import React from 'react';
import RangeSlider from '../index';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { FILL_COLOR_MAP } from '../../../utils/accessibility';

describe('components/RangeSlider', () => {
  let component;
  let handlerObj;

  describe('default fill color', () => {
    beforeEach(() => {
      handlerObj = {
        onChange: () => {},
      };
      spyOn(handlerObj, 'onChange');
      component = shallow(
        <RangeSlider
          value={ 50 }
          onChange={ handlerObj.onChange }
          testSection="myTestSection"
        />
      );
    });

    it('should render component same as snapshot', () => {
      expect(shallowToJson(component)).toMatchSnapshot();
    });

    it('should render component with correct props', () => {
      expect(component.length).toBe(1);
    });

    it('should set the input value from the value prop', () => {
      const input = component.find('input');
      expect(input.prop('value')).toBe(50);
    });

    it('should call the onChange prop when the input onChange is called', () => {
      component.find('input').prop('onChange')();
      expect(handlerObj.onChange).toHaveBeenCalled();
    });

    it('should set a data-test-section attribute from the testSection prop', () => {
      const testSection = component.find('[data-test-section="myTestSection"]');
      expect(testSection.length).toBe(1);
    });
  });

  describe('custom fill color', () => {
    beforeEach(() => {
      handlerObj = {
        onChange: () => {},
      };
      spyOn(handlerObj, 'onChange');
      component = shallow(
        <RangeSlider
          fillColorName="red"
          value={ 50 }
          onChange={ handlerObj.onChange }
          testSection="myTestSection"
          useOffForLabel={ true }
        />
      );
    });

    it('should render the SVG with the fill color that corresponds to the provided fill color name', () => {
      const svgPath = component.find('path').first();
      expect(svgPath.prop('fill')).toBe(`${FILL_COLOR_MAP['red']}`);
    });

    it('should use Off for the label when useOffForLabel is true', () => {
      expect(component.find('.oui-grid').containsMatchingElement(<label>Off</label>)).toBe(true);
    });
  });
});
