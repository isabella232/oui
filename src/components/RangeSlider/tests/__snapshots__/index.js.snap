// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/RangeSlider default fill color should render component same as snapshot 1`] = `
<div
  className="oui-rangeslider"
  data-test-section="myTestSection"
>
  <div
    className="oui-grid"
  >
    <div
      className="oui-grid__cell position--relative"
    >
      <div
        className="range-labels flex"
      >
        <label
          className="oui-label muted flush flex--1"
        >
          0%
        </label>
        <label
          className="oui-label muted flush"
        >
          100%
        </label>
      </div>
      <div
        className="range-display"
      >
        <svg
          height="100"
          preserveAspectRatio="none"
          viewBox="0 0 100 100"
          width="100"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 100h100V0z"
            fill="hsla(227, 100%, 50%, 1)"
            shapeRendering="geometricPrecision"
          />
        </svg>
        <svg
          height="100"
          preserveAspectRatio="none"
          style={
            Object {
              "left": 0,
              "position": "absolute",
              "top": 0,
            }
          }
          viewBox="0 0 100 100"
          width="100"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M50 100H100V0L50 49z"
            fill="white"
            opacity="0.9"
            shapeRendering="geometricPrecision"
          />
        </svg>
      </div>
      <div
        className="range-grid"
      >
        <div
          className="gridline grid-1"
        />
        <div
          className="gridline grid-2"
        />
        <div
          className="gridline grid-3"
        />
        <div
          className="gridline grid-4"
        />
        <div
          className="gridline grid-5"
        />
        <div
          className="gridline gridline-half grid-h1"
        />
        <div
          className="gridline gridline-half grid-h2"
        />
        <div
          className="gridline gridline-half grid-h3"
        />
        <div
          className="gridline gridline-half grid-h4"
        />
      </div>
      <input
        className="hard position--relative width--1-1"
        data-traffic-group="1"
        disabled={false}
        max="100"
        min="0"
        onChange={[Function]}
        step="1"
        type="range"
        value={50}
      />
    </div>
  </div>
</div>
`;
